Correction of Outside Face Conductive Flux Output Variable

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #10089 
 - From a post on unmet hours, a user noticed that the heat stored in a ceiling surface was increasing consistently.  The heat storage was measured by adding the inside and outside conductive flux together at each time step.  Over time, the value should rise and fall, not increase without end.  The problem was traced back to the fact that the surface was an interzone partition and the outside flux was not being calculated.  A switch in the code was skipping all partitions (regular and interzone).  It makes sense to skip the outside calculation for a regular partition because there is technically no "outside".  However, an interzone partition does have an outside and there should be a non-zero conductive flux.  This has been fixed and the user's input file now shows that the stored energy in the interzone surface energy level goes up and down rather than increase non-stop.  Due to the fact that an output variable changed, there are four files where the output has changed.  All four have interzone surfaces and request either the outside flux or the average flux (which is the sum of the inside and outside flux).  So, these changes in the output are expected and should take place.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction to Reporting Problem

The code was skipping partitions for updating the outside face flux in the heat balance because there is no outside face--the surface is adiatic.  However, the code to skip that forgot that interzone partitions do need to figure this out.  This was noticed by a user and the code before the fix resulted in output that made it look like the surface was accumulating energy gain/loss for interzone surfaces over an annual simulation eventhough this value should be relatively small.  With the fix, the outside face flux is now no longer zero all the time and the summation of heat storage in the surface is more appropriate as well.

RKStrand committed:
Outside Face Variable Unit Test and Docs

Unit test added for the outside face flux reporting problem for interzone partition.  Also modified the documentation slightly in the IO Ref to clarify what happens when there are partitions and interzone partitions.

EnergyArchmage commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5175,7 +5175,9 @@ void UpdateThermalHistories(EnergyPlusData &state)
-                if (surface.ExtBoundCond > 0) continue; // Don't need to evaluate outside for partitions
+                // Don't need to evaluate outside for partitions (so continue),
+                // but do need to do remaining calculations for interzone partitions.
+                if (surface.ExtBoundCond > 0 && surface.ExtBoundCond == SurfNum) continue;
```
interesting. This seems closely related to issue #9466, code changes in the same area anyway.   I had found that self-referencing partitions were also not reporting those outside face variables accurately, for similar reasons as here.  But the code I suggested is different.  

`if (surface.ExtBoundCond > 0 && surface.ExtBoundCond != SurfNum) `

perhaps need to get rid of the entire if -> continue, turns out the calculations can't be saved without reporting errors.

Myoldmopar replied:
@EnergyArchmage do you think it's worth talking about this on tomorrow's call?

RKStrand replied:
Sorry--I guess I never responded to @EnergyArchmage here.  My reason for skipping the partitions (when ExtBoundCond = SurfNum) is that the "outside" is really the centerline of those surfaces.  So, the outside flux should be zero because it is adiabatic.  That isn't the case for an interzone partition.  If the partitions also assign the outside surface, I'm not even sure what it will report because there is no outside surface solution for a partition.  That's why I felt like the correct solution was to do what I did.  What were the reporting errors that you noticed?

EnergyArchmage replied:
The output variables Surface Outside Face Conduction Heat Transfer Rate  and Surface Average Face Conduction Heat Transfer Rate never get calculated and Surface Heat Storage Rate has erroneous values.  

I think just remove the if -> continue line altogether.  I believe I put that in way back to save calculation time but it was a mistake.

RKStrand replied:
Ok, I can remove the if -> continue line.  That's not a problem (may have to adjust the unit test--can't remember as it's been a while).  Are you saying that these output variables are not valid and should be eliminated?  Or is there some other clarification that is needed for the IO Ref?

EnergyArchmage replied:
I think the output variables for outside face are useful. It is just that the underlying calcs need to happen for each surface.

RKStrand replied:
Ok, I'll work to eliminate that line (and the new comments that went with it) and then make sure the unit tests are still doing what they need to do.  I'll post here again once I've finished all that and it's ready for another look.

RKStrand replied:
@EnergyArchmage @Myoldmopar Ok, I've done as suggested, fixed the unit test, and clang-formatted everything.  This was just pushed up to GitHub.  Hopefully everything comes back as expected and this is good to go.

Myoldmopar commented on tst/EnergyPlus/unit/HeatBalanceSurfaceManager.unit.cc:
```diff
@@ -8552,4 +8552,79 @@ TEST_F(EnergyPlusFixture, HeatBalanceSurfaceManager_TestUpdateVariableAbsorptanc
+    EXPECT_EQ(15.0, state->dataHeatBalSurf->SurfOpaqInsFaceCondFlux(1));
+    EXPECT_EQ(-15.0, state->dataHeatBalSurf->SurfOpaqOutFaceCondFlux(1));
+    EXPECT_EQ(-15.0, state->dataHeatBalSurf->SurfOpaqInsFaceCondFlux(2));
+    EXPECT_EQ(15.0, state->dataHeatBalSurf->SurfOpaqOutFaceCondFlux(2));
```
😍 I love unit tests that aren't littered with inputs, and instead just set up the necessary structures and do the tests.

Myoldmopar reviewed (commented):
OK, so I'm going to withhold approval for a moment in case we need to discuss this further on the next technicalities call.  Not saying it's wrong, just wanting to pause a second.

nrel-bot-2c commented:
@RKStrand it has been 7 days since this pull request was last updated.

RKStrand committed:
Merge branch 'develop' into 10089OutsideFaceFluxReportingError

RKStrand committed:
Address 10089 Review Comments

Followed @energyarchmage suggestion and got rid of the if->continue.  Changed unit test since the result in it was modified.

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar commented:
Conflict here was just the addition of unit tests in the same spot in multiple branches.  I merged develop in, and resolved it by simply keeping all the unit tests.  Everything builds and tests happily.  I'll let CI do a quick run on it while I break for dinner, but this will merge later if it's happy.

Myoldmopar commented:
Diffs would certainly be expected in the files that use that output variable.  Otherwise all clean.  Merging, thanks @RKStrand
